[cppia] Add tests for missing retyping in function calls

tobil4sk opened the pull request:
This fixes #9990, which crashes due to a missing `TODATAARRAY` instruction, which is required to convert the resulting dynamic array back into an int array.

Before:
```
3	30				CALLSUPER 1  1 1 
3	30					CALLSTATIC 5  6 1 
3	30						TODYNARRAY 
3	30							VAR 6681
```
vs after:
```
3	30				CALLSUPER 1  1 1 
3	30					TODATAARRAY 2 
3	30						CALLSTATIC 5  6 1 
3	30							TODYNARRAY 
3	30								VAR 6681
```




I imagine the same issue can occur with other types of calls, so I will look into that before marking this as ready.

Some of those `(* TODO - proper re-typing *)` comments also look a bit worrying...

tobil4sk commented:
Indeed, we have the same error with super constructor calls, which requires are not covered by the current patch:

```haxe
// template function converts to dynamic array
function wrap<T>(a:Array<T>):Array<T> {
	return a;
}

class Parent {
	public function new(a:Array<Int>) {
		a.copy();
	}
}

class Child extends Parent {
	public function new(a:Array<Int>) {
		super(wrap(a));
	}
}

function main() {
	new Child([1,2,3]);
}
```

@Aidan63 I'm hoping you might have some insight into the cpp generator since you refactored it not long ago. Any chance you might know what the best solution here is? I'm confused with the difference between `CppFunction ((FuncSuperConstruct ...))` and `CppSuper`, or why the former is not currently handled... :sweat_smile:

Aidan63 commented:
Some of the retypers logic can be a bit difficult to follow.

`CppSuper` is any occurance of `super`, the `TField` retyping deals with `CppSuper` by turning them all into `CppFunction FuncSuper` as if the left hand side of a TField expression is `super` it can't be a super constructor call so can safely be considered a class member function super call. 
When retyping `TCall` if it comes across a `CppSuper` then it it can safely assume it's a super constructor since any member function supers will have been retyped by this point using the above logic, so they become `CppFunction FuncSuperConstructor`.

`CppFunction` is any reference to a function, e.g. `final c = foo` where `foo` is a function would result in the right hand side of the assignment operator become a `CppFunction`. You can't do anything with super other than directly call it so any `CppFunction` involving supers will be with a `CppCall`.

I interpreted those "todo proper retyping" comments to refer to the args being retyped to `TCppDynamic` instead of using the arguments type, I've changed that as part of this PR https://github.com/HaxeFoundation/haxe/pull/11981/files#diff-78f353b143a52df58c7c9980eb2f3801f33f84484509ed02009a448182a3b8a4L735

There could well be some missing conversions for cppia in here, I can take a closer look if need be.

tobil4sk committed:
[tests] Add tests for cppia retyping issues

These were solved in:
https://github.com/HaxeFoundation/haxe/pull/11981/changes/4714cc32718fabf8a923efe2fa4578784c1c1319

The following cases from that patch do not affect cppia:
- FuncFromStaticFunction - this is not supported no cppia
- is_array_splice_call - this case has the explicit `not forCppia`
condition
- _hx_create_array_length - this is not used directly on cppia
- cpp_is_templated_call - templated calls are not supported on cppia

There are also some cases where I have added tests, but they were
working correctly before the patch:
- testMapSet - technically the patch makes a difference with regards to
the values stored in the map, but using map.get always converts to the
correct type anyway.
- testDynamicFieldCall - cppia always converts dynamic method arguments
to the correct type. See:
https://github.com/HaxeFoundation/hxcpp/blob/8600e4d49852a3e915c32f97c249ff7111936103/src/hx/cppia/CppiaFunction.cpp#L821
- testObjectDeclaration - similar to map set, it affects the stored data
but the conversion always occurs on dynamic field access.

tobil4sk committed:
[cpp] Remove TODO comments about retyping

tobil4sk commented:
@Aidan63 I went through each line in your patch and added a test for each case that affected cppia with this array conversion issue. This includes #9990, and some other similar cases.

As I mentioned in the commit message, some lines don't affect cppia, and some don't have an observable difference because the `Array<Dynamic>` to `Array<Int>` conversion happens anyway somewhere else.

tobil4sk committed:
Use final consistently

tobil4sk committed:
Remove useless comment

tobil4sk committed:
Remove unused import

tobil4sk commented:
@Simn would you mind also merging this? It just adds a test for an issue that has already been fixed in #11981.
